nixos/caddy: change `acmeCA` default to `null`, omit empty `bind` directive

emilylange opened the pull request:
###### Description of changes

The nixos/caddy module is somewhat old by now and has undergone quite some refactors.
This specific module option (originally named `ca`) used to make a bit more sense when Caddy did not have multiple ACME CAs as fallback (LE & ZeroSSL) by configured by default yet (ZeroSSL came with v2.3.0).

I also rephrased the description slightly, to mention Caddy's automatic issuer fallback and a note which this option maps to in the Caddyfile, to provide a bit more context and a more up-to-date recommendation.

Specifically that "fine-grained configuration" section comes from a time when this module did some custom tls/issuer config json merging with the templated Caddyfile using `jq`.

The "The URL to the ACME CA's directory" section is a word-for-word copy from the official Caddy docs, which also include a link to LE's docs to the referenced staging endpoint. So I added that as well.

#16969, #133671, #147973

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

jian-lin commented on nixos/doc/manual/release-notes/rl-2305.section.md:
```diff
@@ -99,6 +99,8 @@ In addition to numerous new and upgraded packages, this release has the followin
 
 - The [services.unifi-video.openFirewall](#opt-services.unifi-video.openFirewall) module option default value has been changed from `true` to `false`. You will need to explicitly set this option to `true`, or configure your firewall.
 
+- The [services.caddy.acmeCA](#opt-services.caddy.acmeCA) module option default value has been changed from `"https://acme-v02.api.letsencrypt.org/directory"` to the special value `null`, to use all of Caddy's default ACME CA's and enable Caddy's automatic issuer fallback feature by default, as recommended.
```
```suggestion
- The [services.caddy.acmeCA](#opt-services.caddy.acmeCA) module option default value has been changed from `"https://acme-v02.api.letsencrypt.org/directory"` to the special value `null`, to use all of Caddy's default ACME CAs and enable Caddy's automatic issuer fallback feature by default, as recommended.
```

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/1948

SuperSandro2000 commented on nixos/doc/manual/release-notes/rl-2305.section.md:
```diff
@@ -99,6 +99,8 @@ In addition to numerous new and upgraded packages, this release has the followin
 
 - The [services.unifi-video.openFirewall](#opt-services.unifi-video.openFirewall) module option default value has been changed from `true` to `false`. You will need to explicitly set this option to `true`, or configure your firewall.
 
+- The [services.caddy.acmeCA](#opt-services.caddy.acmeCA) module option default value has been changed from `"https://acme-v02.api.letsencrypt.org/directory"` to the special value `null`, to use all of Caddy's default ACME CAs and enable Caddy's automatic issuer fallback feature by default, as recommended.
```
```suggestion
- The [services.caddy.acmeCA](#opt-services.caddy.acmeCA) option now default changed from `"https://acme-v02.api.letsencrypt.org/directory"` to `null`, to use all of Caddy's default ACME CAs and enable Caddy's automatic issuer fallback feature by default, as recommended by upstream.
```

SuperSandro2000 commented on nixos/modules/services/web-servers/caddy/default.nix:
```diff
@@ -245,15 +245,22 @@ in
     acmeCA = mkOption {
-      default = "https://acme-v02.api.letsencrypt.org/directory";
+      default = null;
       example = "https://acme-staging-v02.api.letsencrypt.org/directory";
```
We should adapt the example

IndeedNotJames replied:
To what?

06kellyjac replied:
I'm guessing maybe the testing/staging endpoint for zerossl but IDK if that actually exists

https://caddyserver.com/docs/caddyfile/options#acme-ca

> It is strongly recommended to set this to Let's Encrypt's [staging endpoint](https://letsencrypt.org/docs/staging-environment/) for testing or development. Default: ZeroSSL and Let's Encrypt's production endpoints.

IndeedNotJames replied:
@SuperSandro2000 can you please share what you had in mind?

SuperSandro2000 replied:
We should change the example/description to mention those two endpoints, otherwise we loose information here.

06kellyjac commented:
I'm thinking there could be consequences for people already using caddy via the module changing their CA
Is it worth having a nixos state check and existing deployments use the old value?

SuperSandro2000 commented on nixos/modules/services/web-servers/caddy/default.nix:
```diff
@@ -245,15 +245,22 @@ in
 
-        Set it to `null` if you want to write a more
-        fine-grained configuration manually.
+        Special value `null` should be prefered for production setups,
```
```suggestion
        A value `null` should be prefered for production setups,
```

IndeedNotJames replied:
I wanted to get the point across that this is a nixos-specific option value.
Not a Caddyfile value, considering the nixos module passes a lot of things directly to the resulting Caddyfile.

I could phrase it more clearly, sure.
But replacing "Special" with "a" feels like the opposite direction to me.

IndeedNotJames commented:
@06kellyjac
> I'm thinking there could be consequences for people already using caddy via the module changing their CA
> Is it worth having a nixos state check and existing deployments use the old value?

Sorry for responding that late.
I would argue this isn't really a breaking change.

Caddy has two CAs right now, by default. Let's Encrypt (LE) and ZeroSSL.
LE will be tried first (unless Caddy already knows that rate-limits have been reached) and only then fallback to ZeroSSL.

Sure, both CAs have a different trust-chains.
But this doesn't really matter these days.

Except WebRTC/TURN on Android and iOS (e.g. matrix.org/element.io), where LE isn't trusted, but ZeroSSL is lol

I wrote about that briefly a few months back in https://caddy.community/t/need-help-modifying-yaml-config-to-use-preferred-chains-with-isrg-root-x1/18634/5)

Also, domains/vhosts that leverage [`CAA` dns records](https://en.wikipedia.org/wiki/DNS_Certification_Authority_Authorization) aren't negatively affected by this change either.
E.g. a domain that has `CAA` set up to only allow LE, won't be able to issue a certificate from ZeroSSL.

In such case, Caddy will try LE first, which could fail if LE is down or having issues or whatever, then fallback to ZeroSSL, which fails as well due `CAA` and then just loop around to retry LE.

---

So no, I don't think a `system.stateVersion` is needed here :)

IndeedNotJames commented:
@ofborg test caddy

emilylange committed:
nixos/caddy: change `acmeCA` default to `null`

The nixos/caddy module is somewhat old by now
and has undergone quite some refactors.
This specific module option (originally named
`ca`) used to make a bit more sense when
Caddy did not have multiple ACME CAs as
fallback (LE & ZeroSSL) by configured by
default yet (ZeroSSL came with v2.3.0).

I also rephrased the description slightly,
to mention Caddy's automatic issuer fallback
and a note which this option maps to in the
Caddyfile, to provide a bit more context and
a more up-to-date recommendation.

Specifically that "fine-grained configuration"
section comes from a time when this module did
some custom tls/issuer config json merging
with the templated Caddyfile using `jq`.

The "The URL to the ACME CA's directory"
section is a word-for-word copy from the
official Caddy docs, which also include a link
to LE's docs to the referenced staging
endpoint. So I added that as well.

emilylange committed:
nixos/caddy: omit empty `bind` directive when `listenAddresses` is empty

context: https://caddy.community/t/random-502-errors-when-proxying-to-application-with-self-signed-certificate/19657/6

emilylange commented:
I moved the release note entry from `23.05` to `23.11`.
Would appreciate it, if we could get this into `23.11` :^)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/2302
